### ISSUE TITLE
Can't find hdf5 after apt-get installation

### ISSUE THREAD

**Issue by Pallieguy** [ncbi/ncbi-vdb#19]
Hello,

I'm trying to setup vdb on Ubuntu 14.04 and ran into the following:
```
make install
Checking make status of object libraries...
NOTE - library libkdf5 cannot be built: It requires 'libhdf5' and its development headers.
Installing libraries to /usr/local/ncbi/ncbi-vdb/lib64
Installing includes to /usr/local/ncbi/ncbi-vdb/include
Installing configuration files to /etc/ncbi
Updating /usr/include/ncbi-vdb
Updating /etc/profile.d/ncbi-vdb.[c]sh
Use $NCBI_VDB_LIBDIR in your link commands, e.g.:
      ld -L$NCBI_VDB_LIBDIR -lncbi-vdb ...
Please add /usr/local/ncbi/ncbi-vdb/lib64 to your LD_LIBRARY_PATH, e.g.:
      export LD_LIBRARY_PATH=/usr/local/ncbi/ncbi-vdb/lib64:$LD_LIBRARY_PATH
Use /usr/local/ncbi/ncbi-vdb/lib64 in your link commands, e.g.:
      export NCBI_VDB_LIBDIR=/usr/local/ncbi/ncbi-vdb/lib64
      ld -L$NCBI_VDB_LIBDIR -lncbi-vdb ...
```
Going back to the configure step yields the following
```
./configure
Configuring NCBI-VDB package
checking system type... Linux
checking OS distributor... Ubuntu
checking machine architecture... x86_64
checking for supported architecture... x86_64 (64 bits) is supported
checking for supported OS... Linux (linux) is supported
checking for supported tool chain... gcc tool chain is supported
checking for g++... yes
checking whether gcc accepts -Wno-array-bounds... yes
checking whether g++ accepts -static-libstdc++... no
checking for fuse library... yes
checking for hdf5 library... no
checking for magic library... yes
checking for xml2 library... yes
checking for ngs-sdk package...
	includes... ../ngs/ngs-sdk
	includes... /usr/local/ngs/ngs-sdk
	libraries... /usr/local/ngs/ngs-sdk/lib64
includes: /Apps/ngs/ngs-sdk
libraries: /usr/local/ngs/ngs-sdk/lib64
checking for ngs-java package...
	jar... /usr/local/ngs/ngs-java/jar/ngs-java.jar
jar: /usr/local/ngs/ngs-java/jar/ngs-java.jar.2.9.2

configure: creating 'build/ld.linux.exe_cmd.sh'
configure: creating 'build/Makefile.config.linux.x86_64'
configure: creating 'Makefile.config'
configure: creating 'reconfigure'
build type: release
build prefix: /root/ncbi-outdir/ncbi-vdb
build output path: /root/ncbi-outdir/ncbi-vdb/linux/gcc/x86_64/rel
includedir: /usr/local/ncbi/ncbi-vdb/include
bindir: /usr/local/ncbi/ncbi-vdb/bin
libdir: /usr/local/ncbi/ncbi-vdb/lib
CC = gcc -c
CPP = g++
configured with: ""
```
However I have hdf5-dev installed:
```
apt-get install libhdf5-dev libhdf5-serial-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libhdf5-dev is already the newest version.
libhdf5-serial-dev is already the newest version.
0 upgraded, 0 newly installed, 0 to remove and 162 not upgraded.
```
To verify:
```
ld -lhdf5 --verbose
...
==================================================
attempt to open /usr/x86_64-linux-gnu/lib64/libhdf5.so failed
attempt to open /usr/x86_64-linux-gnu/lib64/libhdf5.a failed
attempt to open //usr/local/lib/x86_64-linux-gnu/libhdf5.so failed
attempt to open //usr/local/lib/x86_64-linux-gnu/libhdf5.a failed
attempt to open //usr/local/lib64/libhdf5.so failed
attempt to open //usr/local/lib64/libhdf5.a failed
attempt to open //lib/x86_64-linux-gnu/libhdf5.so failed
attempt to open //lib/x86_64-linux-gnu/libhdf5.a failed
attempt to open //lib64/libhdf5.so failed
attempt to open //lib64/libhdf5.a failed
attempt to open //usr/lib/x86_64-linux-gnu/libhdf5.so succeeded
-lhdf5 (//usr/lib/x86_64-linux-gnu/libhdf5.so)
libpthread.so.0 needed by //usr/lib/x86_64-linux-gnu/libhdf5.so
found libpthread.so.0 at //lib/x86_64-linux-gnu/libpthread.so.0
libz.so.1 needed by //usr/lib/x86_64-linux-gnu/libhdf5.so
found libz.so.1 at //lib/x86_64-linux-gnu/libz.so.1
libdl.so.2 needed by //usr/lib/x86_64-linux-gnu/libhdf5.so
found libdl.so.2 at //lib/x86_64-linux-gnu/libdl.so.2
libm.so.6 needed by //usr/lib/x86_64-linux-gnu/libhdf5.so
found libm.so.6 at //lib/x86_64-linux-gnu/libm.so.6
libc.so.6 needed by //usr/lib/x86_64-linux-gnu/libhdf5.so
found libc.so.6 at //lib/x86_64-linux-gnu/libc.so.6
ld-linux-x86-64.so.2 needed by //lib/x86_64-linux-gnu/libpthread.so.0
found ld-linux-x86-64.so.2 at //lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
```
I retried after exporting /usr/lib/x86_64-linux-gnu to LD_LIBRARY_PATH, as I read it is a non-standard library location, yet the messages are identical.
Sorry for the edits, I didn't preview before posting.

**Comment by SethM55** [comment]
I was under the impression apt installed the static libraries as well.  There is a libhdf5.a in the same directory.

**Comment by rboissy** [comment]
Here's a summary of a solution to the same problem on Fedora 29.

TL;DR You need to link the 'szip' library explicitly when compiling 'ncbi-vdb' 
and 'sra-tools'.

HDF5's libraries are often linked to the 'szip' compression library, even 
though 'szip' is not OSS:

https://support.hdfgroup.org/doc_resource/SZIP/

Presumably due to the proprietary nature of 'szip', F29's HDF5 libs don't link 
to 'libszip' (aka libsz) and hence their respective rpms don't have a 
dependency on an rpm that provides it. 

https://fedoraproject.org/wiki/Extras/szip

However, as stated on the web page linked to above, the F29 rpm 'libaec' 
provides the shared 'libsz.so', and the F29 rpm 'libaec-devel' provides it 
and the header 'szlib.h':

repoquery -lq --installed libaec-devel
/usr/include/libaec.h
/usr/include/szlib.h
/usr/lib64/libaec.so
/usr/lib64/libsz.so

I'm assuming that on Ubuntu, the 'szip' lib is available, but note that it 
may be named 'libszip'; e.g., the HDF Group names it that way:

ls -1 HDF_Group/HDF5/1.10.4/{include,lib} | grep sz
szip_adpt.h
szlib.h
libszip.a
libszip.so
libszip.so.2
libszip.so.2.1

Here's the relevant excerpts from my F29 installation script for 'sra-tools' 
that might be adaptable to Ubuntu:

ccHdf5='-Wl,-Bstatic -lhdf5 -Wl,-Bdynamic -ldl -lm -lz'

When compiling 'ncbi-vdb':

Prior to running the 'configure' script, and in the same directory as that 
script, run this 'sed' command:

sed -i "s/${ccHdf5}/${ccHdf5} -lsz/" 'setup/konfigure.perl'

When compiling 'sra-tools':

Prior to running the 'configure' script, and in the same directory as that 
script, run these 'sed' commands:

sed -i "s/${ccHdf5}/${ccHdf5} -lsz/" 'setup/konfigure.perl'
sed -i "s/-lpthread/-lpthread -lsz/" 'build/ld.linux.dlib.sh'
sed -i "s/-lpthread/-lpthread -lsz/" 'build/ld.linux.exe.sh'

In all of the above 'sed' commands, the '-lsz' substring may need to be changed 
to '-lszip' as per my previous comments.

Also, the only 'sra-tools' program that seems to be affected by the HDF5/szip 
issue is 'pacbio-load'.

Caveats: I haven't run all of the tests that the NCBI provides for 'sra-tools' 
and haven't used 'pacbio-load' in production.

**Comment by kwrodarmer** [comment]
The message in our configuration script may be misleading. We inform you that a part of VDB won't be built, which was intended to be informational. It does **not** say there was an error or that it was a problem. We **expect** that part of the system to be absent, and in fact, it is only used for a single NCBI-internal tool that you don't need to worry about.

We should probably change the text of the message to say something like "we're not building against hdf5, which is okay because you almost certainly don't need it, but didn't want you to be surprised to find that a submodule of VDB was being left out. It's intentional." Or maybe that's a bit wordy...

**Comment by ayamadanu** [comment]
In my case (CentOS7), I could escape the error by modifying its Makefile as follows (deleted several lines):

# ===========================================================================
#
#                            PUBLIC DOMAIN NOTICE
#               National Center for Biotechnology Information
#
#  This software/database is a "United States Government Work" under the
#  terms of the United States Copyright Act.  It was written as part of
#  the author's official duties as a United States Government employee and
#  thus cannot be copyrighted.  This software/database is freely available
#  to the public for use. The National Library of Medicine and the U.S.
#  Government have not placed any restriction on its use or reproduction.
#
#  Although all reasonable efforts have been taken to ensure the accuracy
#  and reliability of the software and data, the NLM and the U.S.
#  Government do not and cannot warrant the performance or results that
#  may be obtained by using this software or data. The NLM and the U.S.
#  Government disclaim all warranties, express or implied, including
#  warranties of performance, merchantability or fitness for any particular
#  purpose.
#
#  Please cite the author in any work or product based on this material.
#
# ===========================================================================

default: all

TOP ?= $(abspath ../..)
MODULE = libs/hdf5

INT_LIBS = \

EXT_LIBS = \
	libkdf5

ALL_LIBS = \
	$(INT_LIBS) \
	$(EXT_LIBS)

OBJFILES = \
	$(KDF5_OBJ)

include $(TOP)/build/Makefile.env

ifdef HDF5_INCDIR
	INCDIRS += $(addprefix -I,$(HDF5_INCDIR))
endif

#-------------------------------------------------------------------------------
# outer targets
#
all std: makedirs compile
	@ $(MAKE_CMD) $(TARGDIR)/std

$(INT_LIBS): makedirs
	@ $(MAKE_CMD) $(ILIBDIR)/$@

$(EXT_LIBS): makedirs
	@ $(MAKE_CMD) $(LIBDIR)/$@

.PHONY: all std $(ALL_LIBS)

#-------------------------------------------------------------------------------
# std
#
$(TARGDIR)/std: \
	$(addprefix $(ILIBDIR)/,$(INT_LIBS)) \
	$(addprefix $(LIBDIR)/,$(EXT_LIBS))

.PHONY: $(TARGDIR)/std

#-------------------------------------------------------------------------------
# clean
#
clean: stdclean

.PHONY: clean

#-------------------------------------------------------------------------------
# hdf5
#
$(LIBDIR)/libkdf5: $(addprefix $(LIBDIR)/libkdf5.,$(LIBEXT))

KDF5_SRC = \
	hdf5dir \
	hdf5file \
	hdf5arrayfile

KDF5_OBJ = \
	$(addsuffix .$(LOBX),$(KDF5_SRC))

KDF5_LIB = \
	-lhdf5 \
	-lz \

ifdef HDF5_LIBDIR
	KDF5_LIB += $(addprefix -L,$(HDF5_LIBDIR))
endif

$(LIBDIR)/libkdf5.$(SHLX): $(KDF5_OBJ)
	$(LD) --dlib --vers $(SRCDIR)/../ncbi-vdb/libncbi-vdb.vers -o $@ $^ $(KDF5_LIB)

$(LIBDIR)/libkdf5.$(LIBX): $(KDF5_OBJ)
	$(LD) --slib --vers $(SRCDIR)/../ncbi-vdb/libncbi-vdb.vers -o $@ $^ $(KDF5_LIB)

compile: stdcompile

$(TARGDIR)/compile: $(OBJFILES)

**Comment by durbrow** [comment]
**PLEASE NOTE**
You do **not** need HDF5 to use SRA files!

If your intent is to read SRA files, you do not need the functionality at is enabled by HDF5. This functionality is used process data that is submitted to SRA. SRA turns all submitted data into a single format.
